src/viur/core/modules/history.py

```diff
@@ -0,0 +1,742 @@
+                }
+
+            schema = bone_to_schema(name, bone)
+
+            if isinstance(schema, dict):
+                self.schema.append(schema)
+            else:
+                self.schema.extend(schema)
+
+    def _get_table(self, skel):
+        # FIXME: Currently not in use!
+        if self.table:
+            return self.table
+
+        try:
+            self.table = self.client.get_table(self.path)
+
+        except exceptions.NotFound:
+            app, dataset, table = self.path.split(".")
+
+            self.__generate_schema()
+
+            # create dataset if needed
+            try:
+                self.client.get_dataset(dataset)
+            except exceptions.NotFound:
+                self.client.create_dataset(dataset)
+
+            # create table if needed
+            try:
+                self.table = self.client.get_table(self.path)
+            except exceptions.NotFound:
+                self.table = bigquery.Table(
+                    self.path,
+                    schema=self.schema
+                )
+                self.client.create_table(table)
+
+        return self.table
+
+    def write_skel(self, skel):
+        # FIXME: Currently not in use!
+        self._get_table(skel)
+
+        def skel_to_bigquery(skel, prefix="") -> dict:
+            values = {}
+
+            for name, bone in skel.items():
+                if name in self.IGNORE_BONES:
+                    continue
+
+                if isinstance(bone, RelationalBone):
+                    values |= skel_to_bigquery(skel[name], prefix + name + "_")
+                else:
+                    values[prefix + name] = skel[name]
+
+            return values
+
+        return self.client.insert_rows(self.table, [skel_to_bigquery(skel)])
+
+
+class ViurHistory(List):
+    """
+    ViUR history module
+    """
+    kindName = "viur-history"
+
+    adminInfo = {
+        "name": "Protokoll",
+        "icon": "clock-history",
+        "filter": {
+            "orderby": "timestamp",
+            "orderdir": "1",
+        },
+        "disabledActions": ["add", "clone", "delete"],
+    }
+
+    roles = {
+        "admin": "view",
+    }
+
+    HISTORY_VERSION = 1
+    """
+    History format version.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.bigquery = BigQueryHistory()
+
+    def baseSkel(self):
+        # Make all bones readOnly!
+        # FIXME: There should be a skel.readonly() function soon...
+        skel = super().baseSkel().clone()
+
+        for bone in skel.values():
+            bone.readOnly = True
+
+        return skel
+
+    def canEdit(self, skel):
+        return self.canView(skel)
+
+    def canDelete(self, _skel):
+        return False
+
+    def canAdd(self):
+        return False
+
+    # Module-specific functions
+
+    def _create_diff(self, new: dict, old: dict, diff_excludes: {str} = set()):
```

Comment:
Are you sure this is a valid annotation? At least PyCharm does not understand it :/

```suggestion
    def _create_diff(self, new: dict, old: dict, diff_excludes: set[str] = set()):
```